### ISSUE TITLE
go back to typed arrays for indices?

### ISSUE THREAD

**Issue by gordonwoodhull** [crossfilter/crossfilter#134]
@monfera noticed while working on #122 that typed arrays were inadvertently disabled for indices in the refactoring for browserify (70fb3dca5458d3).

It would be interesting to test if performance dropped due to this. Unless modern JS engines are so smart that they can optimize untyped arrays into typed ones, it should help caching to use the typed arrays.

**Comment by monfera** [comment]
Sorting is a process that can't be done entirely in the CPU cache esp. L1 cache unless the size is trivial, which is not the case with the 250k records in the current benchmark. Untyped, homogeneous numerical arrays, I suppose, still store every element as a 64 bit floating point number. 

If the dimension count is low, the original (and pre-2016) version of `crossfilter` would use a byte array, therefore 8 times as many elements would fit into the operational CPU cache levels, increasing bandwidth, reducing latency and maybe allowing the process to only deal with fewer cache levels, though indexing into another, generally unsorted value array for the comparisons will thrash the cache anyway.

So, if the indexing part of the benchmark doesn't improve with the typed array, perf issues could be filed for browser makers, as in this case, maybe the browser's native sorter doesn't directly operate on the low bitcount numbers.

It's likely that the more compact representation translates to even more modest savings when filtering (querying) because filtering and grouping involves a bunch of other data, more indexed access, invocation of reducers, and hopefully, touching a very small set of data points due to its incremental nature.

**Comment by gordonwoodhull** [comment]
Yeah, I expect we'll see significant performance restored with this fixed. 

We did [notice a performance drop from 1.3 to 1.4](https://github.com/crossfilter/crossfilter/releases/tag/1.4.0) but I think most of us attributed it to the >32 dimensions. This seems more likely! Did you ever get it to run, @monfera?

We should really have a line chart of performance over time. It wouldn't be hard to do since the interface hasn't changed much.

**Comment by esjewett** [comment]
Nice find @gordonwoodhull

**Comment by monfera** [comment]
@gordonwoodhull I didn't get around to doing so, the obvious quick attempt
```
module.exports = {
  array8: crossfilter_array8,
  array16: crossfilter_array16,
  array32: crossfilter_array32,
```
yielded an error in `npm run benchmark`:
```
/Users/robert/repos/monfera/crossfilter/src/crossfilter.js:1158
          g.value = reduceAdd(g.value, data[i], true);
                                ^

TypeError: Cannot read property 'value' of undefined
    at resetMany (/Users/robert/repos/monfera/crossfilter/src/crossfilter.js:1158:33)
```
and I've no plans to use cf now, just poking the code to refresh my memory and grok choices - even Mike's 1st commit is a fantastic tour de force, complete with Google Closure Compiler-like manual code compression - I had to circle back to Elastic tasks, got some talk etc. next week.

**Comment by monfera** [comment]
OK here's some wild speculation. Both I and @gordonwoodhull observed that #122 makes filtering was 0..25% faster (average <=10%), depending on dimension etc. It made me think that a carefully optimized built-in sorter may choose to perform a first pass to check the domain of the data, and do various things, eg. balk if the array is already sorted, or... copy the array to an unsigned 8-bit array if it's all integers and they happen to fit in that space, sort that, and only convert the array back to an FP array if some operation requires so (not the case with crossfilter). Nothing stops a JS engine to still pretend for the userland that it's an untyped array, such cool tricks have [prior art](https://github.com/crossfilter/crossfilter/pull/122#issuecomment-496578264). 

If this were the case, subsequent ops on the sorted array could also become a bit faster, assuming it retained its 1/8th size and bit handling friendly ops, more of it fitting in the cache too.

It was going to be a long shot, but as luck would have it, a search pointed to at least _some_ evidence that the V8 engine is not at all alien to this type of thinking, even for sorting specifically 😄   see the link to `ElementsKind` [here](https://v8.dev/blog/array-sort#performance-trade-offs). I've seen several userland sorters in JS libs that worked on typed, narrower arrays so the V8 guys are even more likely to test for such obvious cache friendliness (again, sorting is non-local so fitting multiple times as many elements in the cache can be a big win).

One way to test it is to mix random FP operations which might (or might not) undo the internal 8-bit representation; another is to see in non-V8 browsers how performance evolves with/without the PR, once `master` is back to a typed array.

So it's possible that reverting to typed arrays will yield about the same _filtering_ performance increase as would the PR to switch to the built-in sort; it's possible that the type-fixed `master` will be more performant in browsers with less optimized built-in `sort` than my PR. So perhaps even with 7 years of browser improvements since, a userland sort can be worth the extra 400 lines.

**Comment by monfera** [comment]
Whoa search for `PACKED_SMI_ELEMENTS` here: https://v8.dev/blog/elements-kinds - small integers!

**Comment by gordonwoodhull** [comment]
I spoke too soon. I am seeing a small performance improvement with typed arrays restored, but it is not as big as I hoped.

I think @monfera is right that V8 is able to detect that the arrays should be typed, and probably the integer width as well. So explicitly using typed arrays just pushes V8 in the right direction earlier.

My branch is here: https://github.com/gordonwoodhull/crossfilter/tree/restore-typed-arrays

I think I saw a more significant performance improvement for Firefox, but I want to do more testing. I created a very hacky version of the benchmark for browsers but it was too gross to check in. I should clean that up and include it for the github.io site.

I also tried some historical testing of performance. The benchmark goes back to the initial commit, so this is not that hard. But I didn't see any interesting results yet, so I probably got something wrong. It would be reassuring to see which commits hurt/help performance.

**Comment by monfera** [comment]
We should crossfilter the perf results :-) Dimension 1: commit time. Dimension 2: browser. Dimension 3: Browser vintage (browser perf itself varies over time). Dimension 4: platform (desktop, laptop, top phone, average phone, low end phone, OS X vs Android) Dimension 5: Time of hardware generation. So it could be a multitemporal query of itself 😂

**Comment by gordonwoodhull** [comment]
It turned out there were a couple of more recent changes which didn’t work with typed arrays — described in #145 and fixed in 6a8e1741e56 and b8d867868d.

I didn’t attempt to benchmark this (or build a data collecting per-user-agent profiling page haha!) I expect it will help performance on browsers with weaker JS optimization.
